exchanges/okx/okx.go

```diff
@@ -0,0 +1,4847 @@
+	}
+	input := map[string][]string{"ccy": currency}
+	var resp response
+	er := ok.SendHTTPRequest(ctx, exchange.RestSpot, smallAssetsConvertEPL, http.MethodPost, smallAssetsConvert, input, &resp, true)
+	if er != nil {
+		return nil, er
+	}
+	if len(resp.Data) == 0 {
+		if resp.Msg != "" {
+			return nil, errors.New(resp.Msg)
+		}
+		return nil, errNoValidResponseFromServer
+	}
+	return &resp.Data[0], nil
+}
+
+// GetSavingBalance returns saving balance, and only assets in the funding account can be used for saving.
+func (ok *Okx) GetSavingBalance(ctx context.Context, currency string) ([]SavingBalanceResponse, error) {
+	type response struct {
+		Code string                  `json:"code"`
+		Msg  string                  `json:"msg"`
+		Data []SavingBalanceResponse `json:"data"`
+	}
+	var resp response
+	params := url.Values{}
+	if currency != "" {
+		params.Set("ccy", currency)
+	}
+	path := common.EncodeURLValues(assetSavingBalance, params)
+	return resp.Data, ok.SendHTTPRequest(ctx, exchange.RestSpot, getSavingBalanceEPL, http.MethodGet, path, nil, &resp, true)
+}
+
+// SavingsPurchase posts a purchase
+func (ok *Okx) SavingsPurchase(ctx context.Context, arg *SavingsPurchaseRedemptionInput) (*SavingsPurchaseRedemptionResponse, error) {
+	if arg == nil {
+		return nil, errors.New("invalid argument, null argument")
+	}
+	arg.ActionType = "purchase" // purchase: purchase saving shares
+	return ok.savingsPurchaseOrRedemption(ctx, arg)
+}
+
+// SavingsRedemption posts a redemption
+func (ok *Okx) SavingsRedemption(ctx context.Context, arg *SavingsPurchaseRedemptionInput) (*SavingsPurchaseRedemptionResponse, error) {
+	if arg == nil {
+		return nil, errors.New("invalid argument, null argument")
+	}
+	arg.ActionType = "redempt" // redempt: redeem saving shares
+	return ok.savingsPurchaseOrRedemption(ctx, arg)
+}
+
+// savingsPurchaseOrRedemption creates a purchase or redemption instance
+func (ok *Okx) savingsPurchaseOrRedemption(ctx context.Context, arg *SavingsPurchaseRedemptionInput) (*SavingsPurchaseRedemptionResponse, error) {
+	if arg == nil {
+		return nil, errNilArgument
+	}
+	switch {
+	case arg.Currency == "":
+		return nil, errInvalidCurrencyValue
+	case arg.Amount <= 0:
+		return nil, errUnacceptableAmount
+	case !(arg.ActionType == "purchase" || arg.ActionType == "redempt"):
+		return nil, errors.New("invalid side value, side has to be either \"redemption\" or \"purchase\"")
+	case arg.ActionType == "purchase" && !(arg.Rate >= 1 && arg.Rate <= 365):
+		return nil, errors.New("the rate value range is between 1% and 365%")
+	}
+	type response struct {
+		Code string                              `json:"code"`
+		Msg  string                              `json:"msg"`
+		Data []SavingsPurchaseRedemptionResponse `json:"data"`
+	}
+	var resp response
+	er := ok.SendHTTPRequest(ctx, exchange.RestSpot, savingsPurchaseRedemptionEPL, http.MethodPost, assetSavingPurchaseOrRedemptionPath, &arg, &resp, true)
```

Comment:
This looks quite elaborate and instead should just be one function. Where `side` is populated based on the string provided.